### PR TITLE
Removing Microsoft.bcl* packages

### PR DESCRIPTION
**Pull request by Anipik**

### CODE REVIEW

File: src/System.Text.Json/ref/System.Text.Json.csproj [hunk 1]
```diff
@@ -5,17 +5,23 @@
   <ItemGroup>
     <Compile Include="System.Text.Json.cs" />
   </ItemGroup>
-  <ItemGroup Condition="'$(TargetGroup)' != 'netstandard' AND '$(TargetsNetFx)' != 'true'">
+  <ItemGroup Condition="'$(TargetGroup)' == 'netcoreapp' OR '$(TargetGroup)' == 'uap'">
```
Comment (ViktorHofer) [inline src/System.Text.Json/ref/System.Text.Json.csproj]: nit: TargetFramework

File: src/System.Text.Json/ref/System.Text.Json.csproj [hunk 2]
```diff
@@ -5,17 +5,23 @@
...
     <ProjectReference Include="..\..\System.Memory\ref\System.Memory.csproj" />
     <ProjectReference Include="..\..\System.Runtime\ref\System.Runtime.csproj" />
     <ProjectReference Include="..\..\System.Text.Encodings.Web\ref\System.Text.Encodings.Web.csproj" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetGroup)' == 'netcoreapp3.0'">
```
Comment (ViktorHofer) [inline src/System.Text.Json/ref/System.Text.Json.csproj]: nit: TargetFramework

File: src/System.Text.Json/ref/Configurations.props [hunk 3]
```diff
@@ -3,6 +3,7 @@
     <PackageConfigurations>
       netstandard;
       net461;
+      netcoreapp3.0;
```
Comment (ViktorHofer) [inline src/System.Text.Json/ref/Configurations.props]: you want to add a netcoreapp3.0 asset to the System.Text.Json package?

File: eng/restore/binplacePackages/binplacePackages.depproj [hunk 4]
```diff
@@ -33,6 +33,10 @@
       <Version>4.5.3</Version>
     </PackageReference>
 
+    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Condition="('$(TargetsNetStandard)' == 'true' AND '$(NETStandardVersion)' &gt;= 2.0) OR '$(TargetsNetStandard)' != 'true'">
```
Comment (ViktorHofer) [inline eng/restore/binplacePackages/binplacePackages.depproj]: TargetsNetStandard != true OR NS >= 2.0

**Review by ViktorHofer — COMMENTED** [review]
This is a dupe of https://github.com/dotnet/corefx/pull/40685 but I'll close mine.

File: src/System.Text.Json/ref/Configurations.props (hunk 3 above)
Comment (Anipik) [reply src/System.Text.Json/ref/Configurations.props]: yes I will add it to 3.1 branch as well. cc @ericstj can explain the reason better than me

Comment (ericstj) [reply src/System.Text.Json/ref/Configurations.props]: Without this, we bring in the `netstandard2.0` ref on `netcoreapp3.0` which in-turn brings in the `Microsoft.BCL.AsyncInterfaces` package.  We'd like to avoid bringing in this dependency on `netcoreapp3.0`.  /cc @ahsonkhan @joperezr

File: eng/restore/binplacePackages/binplacePackages.depproj (hunk 4 above)
Comment (ViktorHofer) [reply eng/restore/binplacePackages/binplacePackages.depproj]: this doesn't seem right. AFAIK for sourcebuild we can't use pre-built bits on netcoreapp. cc @ericstj

File: src/System.Text.Json/ref/System.Text.Json.csproj (hunk 1 above)
Comment (Anipik) [reply src/System.Text.Json/ref/System.Text.Json.csproj]: Dont we use TargetGroup with versionless and targetFramework with specific versions ?

Comment (ViktorHofer) [reply src/System.Text.Json/ref/System.Text.Json.csproj]: oh right. ignore that one.

File: src/System.Text.Json/ref/Configurations.props (hunk 3 above)
Comment (ericstj) [reply src/System.Text.Json/ref/Configurations.props]: An alternative here could be to target `netstandard2.1` to avoid the dependency.  With that we wouldn't need BuildConfigurations, I think.

File: src/shims/ApiCompatBaseline.netcoreapp.netstandard.txt [hunk 5]
```diff
@@ -67,4 +67,14 @@ MembersMustExist : Member 'System.Linq.EnumerableQuery..ctor()' does not exist i
 MembersMustExist : Member 'System.Threading.Tasks.TaskExtensions.ConfigureAwait(System.IAsyncDisposable, System.Boolean)' does not exist in the implementation but it does exist in the contract.
 MembersMustExist : Member 'System.Threading.Tasks.TaskExtensions.ConfigureAwait<T>(System.Collections.Generic.IAsyncEnumerable<T>, System.Boolean)' does not exist in the implementation but it does exist in the contract.
 MembersMustExist : Member 'System.Threading.Tasks.TaskExtensions.WithCancellation<T>(System.Collections.Generic.IAsyncEnumerable<T>, System.Threading.CancellationToken)' does not exist in the implementation but it does exist in the contract.
-Total Issues: 59
+TypesMustExist : Type 'System.IAsyncDisposable' does not exist in the implementation but it does exist in the contract.
```
Comment (ericstj) [inline src/shims/ApiCompatBaseline.netcoreapp.netstandard.txt]: These are here because we aren't binplacing into `netcoreapp` (versionless).  We could do that too.

File: eng/restore/binplacePackages/Configurations.props [hunk 6]
```diff
@@ -13,6 +13,7 @@
       netstandard;
       netcoreapp2.0;
       netcoreapp3.0;
+      netcoreapp;
```
Comment (ViktorHofer) [inline eng/restore/binplacePackages/Configurations.props]: @ericstj I thought sourcebuild doesn't allow this?

Comment (ericstj) [reply eng/restore/binplacePackages/Configurations.props]: `netstandard` will also be [restored](https://github.com/dotnet/corefx/blob/6549a93fb462f581696af11a3be076d36309dc7e/eng/restore/dirs.proj#L5) for source build.  Source build will need to add [stand-ins](https://github.com/dotnet/source-build-reference-packages) for these packages, as is the case for any of the packages which we dead-end (we won't be the only repo that needs to consume them).

File: src/System.Text.Json/ref/Configurations.props (hunk 3 above)
Comment (joperezr) [reply src/System.Text.Json/ref/Configurations.props]: I see I suppose this is because conflict resolution will find that netstandard2.0 asset has a higher assembly version than inbox? I also guess that we don't want to put a placeholder on netcoreapp3.0 TFM in order to not freeze the API?

Comment (joperezr) [reply src/System.Text.Json/ref/Configurations.props]: Would this cause a problem with roll-forward?

Comment (ericstj) [reply src/System.Text.Json/ref/Configurations.props]: > Would this cause a problem with roll-forward?

Note that we're only changing the ref here, which is not subject to roll-forward.  The src lib already includes the nca3.0 config.

The goal with System.Text.Json is to simultaneously rev package and inbox, similar to S.C.Immutable/Reflection.Metadata.  The reason for the nca3.0 ref assembly is so that we accurately represent the distinct configs of the library.  We were already including the 3.0 lib in the package.  Omitting the 3.0 ref *works* (since there is no surface area unique to nca3.0) but its less than ideal.  Including the 3.0 ref lets us avoid depending on the compat packages just for compile.  We could choose to do ns2.1 instead, but I think it makes more sense to keep parallel configs to the src lib.

**Comment by Anipik** [comment]
@ericstj @ViktorHofer  is there something else that needs to be done here ?

**Comment by ericstj** [comment]
I'd like @ahsonkhan to sign off on the Json change.

**Review by ericstj — APPROVED** [review]

File: src/System.Text.Json/ref/Configurations.props (hunk 3 above)
Comment (ericstj) [reply src/System.Text.Json/ref/Configurations.props]: Actually took a closer look here and I think we should just clean this up.  The reference isn't even shipping inside the package, so my concern about leaking the Microsoft.BCL.* package to netcoreapp3.0 is a non-issue.  Instead of doing this, can we delete the PackageConfigurations (just use BuildConfigurations) and remove this ProjectReference from the pkgproj?  Not sure if we had a reason for doing it the way it is now, or we just missed cleaning this up before.  @joperezr do you recall?

File: src/System.Text.Json/pkg/System.Text.Json.pkgproj [hunk 7]
```diff
@@ -1,9 +1,6 @@
 <Project DefaultTargets="Build">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), Directory.Build.props))\Directory.Build.props" />
   <ItemGroup>
-    <ProjectReference Include="..\ref\System.Text.Json.csproj">
-      <SupportedFramework>net461;netcoreapp2.0;uap10.0.16299;$(AllXamarinFrameworks)</SupportedFramework>
-    </ProjectReference>
     <ProjectReference Include="..\src\System.Text.Json.csproj">
```
Comment (ericstj) [inline src/System.Text.Json/pkg/System.Text.Json.pkgproj]: Please remove the ExcludeReferenceAssets property below, add a comment here indicating why we wish to exclude the reference project.

File: src/System.Text.Json/ref/System.Text.Json.csproj [hunk 8]
```diff
@@ -5,7 +5,7 @@
   <ItemGroup>
     <Compile Include="System.Text.Json.cs" />
   </ItemGroup>
-  <ItemGroup Condition="'$(TargetGroup)' != 'netstandard' AND '$(TargetsNetFx)' != 'true'">
```
Comment (ahsonkhan) [inline src/System.Text.Json/ref/System.Text.Json.csproj]: Is this change necessary now?

It is better for the condition to clearly be a negative of the other (like if/else).
```csproj
<ItemGroup Condition="...NOT X...">
</ItemGroup>

<ItemGroup Condition="...X...">
</ItemGroup>
```

File: src/System.Text.Json/pkg/System.Text.Json.pkgproj (hunk 7 above)
Comment (Anipik) [reply src/System.Text.Json/pkg/System.Text.Json.pkgproj]: The reason being that we dont ship ref inside the nuget package ?

**Comment by Anipik** [comment]
@ericstj @ahsonkhan I addressed the feedback

File: eng/restore/binplacePackages/Configurations.props (hunk 6 above)
Comment (ViktorHofer) [reply eng/restore/binplacePackages/Configurations.props]: Don't you want to revert this change?

Comment (Anipik) [reply eng/restore/binplacePackages/Configurations.props]: yes this should be removed.

File: eng/restore/binplacePackages/Configurations.props [hunk 9]
```diff
@@ -12,6 +12,7 @@
       netstandard2.0;
       netstandard;
       netcoreapp2.0;
+      netcoreapp3.0;
```
Comment (ViktorHofer) [inline eng/restore/binplacePackages/Configurations.props]: why is this still required?

Comment (Anipik) [reply eng/restore/binplacePackages/Configurations.props]: I think we don't require netcoreapp3.0 but require netcoreapp for apicompact between netcoreapp and netstandard. I am just verifying.

Comment (ViktorHofer) [reply eng/restore/binplacePackages/Configurations.props]: We should not restore anything here just for the sake of ApiCompat.

Comment (Anipik) [reply eng/restore/binplacePackages/Configurations.props]: This would result in adding these types ApiCompatBaseline.Txt files which is also not good.
We should not use apiCompatBaseline.txt files to work around stuff. 
The potential harm here is people adding a reference to something that they shouldn't. but this can be caught in reviews.
However, I don't feel strongly either way, just wanted to bring up why adding netcoreapp could be good.

Comment (Anipik) [reply eng/restore/binplacePackages/Configurations.props]: We still require netcoreapp3.0 because we still build the src with config netcoreapp3.0 and it will otherwise pick a netstandard dll for reverseApiCompat

**Comment by Anipik** [comment]
@ericstj @ViktorHofer can you give a final review here ?

File: eng/restore/binplacePackages/Configurations.props (hunk 9 above)
Comment (ViktorHofer) [reply eng/restore/binplacePackages/Configurations.props]: @ericstj I'm still unsure about binplacing restored netcoreapp5.0 assets here. I don't think that's a good idea but I'm happy to be proven wrong. When I created that depproj it was never intended to be used for netcoreapp.

Comment (Anipik) [reply eng/restore/binplacePackages/Configurations.props]: Resolved offline

File: eng/restore/binplacePackages/binplacePackages.depproj [hunk 10]
```diff
@@ -33,6 +33,10 @@
       <Version>4.5.3</Version>
     </PackageReference>
 
+    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Condition="'$(TargetsNetStandard)' != 'true' OR '$(NETStandardVersion)' &gt;= 2.0">
```
Comment (ViktorHofer) [inline eng/restore/binplacePackages/binplacePackages.depproj]: As discussed in the standup, please add a choose when otherwise element in the file that makes sure that nothing gets restored for netcoreapp that is unintended. And please also add comments.

**Comment by Anipik** [comment]
@ViktorHofer @ericstj  I added the choose when statement

File: eng/restore/binplacePackages/binplacePackages.depproj [hunk 11]
```diff
@@ -26,24 +26,40 @@
...
+        <PackageReference Include="System.Numerics.Vectors" Condition="'$(TargetsNetStandardLowerThan21)' == 'true' OR '$(TargetsNetfx)' == 'true'">
+          <Version>4.5.0</Version>
+        </PackageReference>
+
+        <PackageReference Include="System.Threading.Tasks.Extensions" Condition="'$(TargetFramework)' == 'netcoreapp2.0' OR '$(TargetsNetfx)' == 'true' OR '$(TargetsNetStandard)' == 'true'">
+          <Version>4.5.2</Version>
+        </PackageReference>
+
+        <!-- Only include the assets from the direct packages we reference in the output -->
+        <PackageToInclude Include="@(PackageReference)" />
```
Comment (ericstj) [inline eng/restore/binplacePackages/binplacePackages.depproj]: I believe this should be common, move out of choose/when.

File: eng/restore/binplacePackages/binplacePackages.depproj [hunk 12]
```diff
@@ -26,24 +26,40 @@
...
-    <PackageToInclude Include="@(PackageReference)" />
-  </ItemGroup>
+  <Choose>
+    <!-- This is to make sure that nothing unintended gets restored for netcoreapp. -->
+    <When Condition="'$(TargetGroup)' != 'netcoreapp'">
+      <ItemGroup>
+        <!-- runtime dependency: System.Data.SqlClient uap10.0.16299 -->
+        <!-- runtime dependency: System.Diagnostics.PerformanceCounters netcoreapp2.0,net461 -->
+        <PackageReference Include="System.Memory" Condition="'$(TargetFramework)' == 'netcoreapp2.0' OR '$(TargetFramework)' == 'uap10.0.16299' OR '$(TargetsNetfx)' == 'true' OR ('$(TargetsNetStandard)' == 'true' AND '$(NETStandardVersion)' &gt;= 1.1)">
+         <Version>4.5.3</Version>
```
Comment (ericstj) [inline eng/restore/binplacePackages/binplacePackages.depproj]: For brevity can we use attribute syntax instead of child node?

**Comment by Anipik** [comment]
@ericstj is it good to merge this one ?

File: eng/restore/binplacePackages/binplacePackages.depproj (hunk 11 above)
Comment (ericstj) [reply eng/restore/binplacePackages/binplacePackages.depproj]: I meant that PackageToInclude  should be common,.

**Review by ericstj — COMMENTED** [review]
See my reply.

File: eng/restore/binplacePackages/binplacePackages.depproj (hunk 11 above)
Comment (Anipik) [reply eng/restore/binplacePackages/binplacePackages.depproj]: okay. we can still keep the AsyncInterface package in common ?

Comment (ericstj) [reply eng/restore/binplacePackages/binplacePackages.depproj]: You can, but a choose/when with only one when clause is the same thing as a condition on the itemgroup.  Maybe just reduce to that?  Trying to address @ViktorHofer's concern without making this thing too complex to reason about.

File: src/shims/ApiCompatBaseline.uap.netstandard.txt [hunk 13]
```diff
@@ -137,4 +137,14 @@ CannotSealType : Type 'System.Linq.EnumerableExecutor' is effectively (has a pri
 MembersMustExist : Member 'System.Linq.EnumerableExecutor..ctor()' does not exist in the implementation but it does exist in the contract.
 CannotSealType : Type 'System.Linq.EnumerableQuery' is effectively (has a private constructor) sealed in the implementation but not sealed in the contract.
 MembersMustExist : Member 'System.Linq.EnumerableQuery..ctor()' does not exist in the implementation but it does exist in the contract.
-Total Issues: 121
+TypesMustExist : Type 'System.IAsyncDisposable' does not exist in the implementation but it does exist in the contract.
```
Comment (ericstj) [inline src/shims/ApiCompatBaseline.uap.netstandard.txt]: I guess we're missing UAP from the restore project as well, here I care less.

**Review by ericstj — APPROVED** [review]

File: eng/restore/binplacePackages/binplacePackages.depproj [hunk 14]
```diff
@@ -26,21 +26,17 @@
...
-
-    <PackageReference Include="System.Numerics.Vectors" Condition="'$(TargetsNetStandardLowerThan21)' == 'true' OR '$(TargetsNetfx)' == 'true'">
-      <Version>4.5.0</Version>
-    </PackageReference>
-
-    <PackageReference Include="System.Threading.Tasks.Extensions" Condition="'$(TargetFramework)' == 'netcoreapp2.0' OR '$(TargetsNetfx)' == 'true' OR '$(TargetsNetStandard)' == 'true'">
-      <Version>4.5.2</Version>
-    </PackageReference>
-
+    <PackageReference Include="System.Memory" Version="4.5.2" Condition="'$(TargetFramework)' == 'netcoreapp2.0' OR '$(TargetFramework)' == 'uap10.0.16299' OR '$(TargetsNetfx)' == 'true' OR ('$(TargetsNetStandard)' == 'true' AND '$(NETStandardVersion)' &gt;= 1.1)" />
```
Comment (ericstj) [inline eng/restore/binplacePackages/binplacePackages.depproj:33]: Looks like this regressed System.Memory ref, discovered here: https://github.com/dotnet/sdk/pull/3884
